api/src/sparql.rs

```diff
@@ -0,0 +1,63 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//!
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::dataset::Dataset;
+use crate::graph::Graph;
+use crate::prefix::PrefixMap;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset: Dataset {
+    type Result: SparqlResult;
+    type SparqlError: Error + 'static;
+
+    /// Parse and immediately execute `query`
+    fn query<'p, P>(&self, query: &str) -> Result<Self::Result, Self::SparqlError>
+    where
+        P: PrefixMap<'p> + ?Sized;
+}
+
+/// The result of executing a SPARQL query.
+pub trait SparqlResult {
+    type Term: TTerm;
+    /// The type of iterator returned if the query was a SELECT.
+    type Bindings: Iterator<Item = Vec<Option<Self::Term>>>;
+    /// The type of graph returned if the query was a CONSTRUCT or DESCRIBE.
+    type Graph: Graph;
+
+    /// The kind of this result (depends on the type of query)
+    fn kind(&self) -> SparqlResultKind;
+    /// If the query was a SELECT, the list of selected variable names
+    fn variables(&self) -> Option<Vec<String>>;
+    /// If the query was a SELECT, an iterator over all the solutions
+    fn bindings(self) -> Option<Self::Bindings>;
+    /// If the query was an ASK, the boolean result
+    fn boolean(&self) -> Option<bool>;
+    /// If the query was a CONSTRUCT or DESCRIBE, the resulting graph
+    fn graph(&self) -> Option<Self::Graph>;
```

Comment:
Taking  `&self` and not `self` might requires most of implementations to copy data here because Self::Graph lifetime is not limited by `&self` lifetime.

Reply:
You are right :-/

I am tempted to replace `&self` by `self`. Or do you suggest another option?

Reply:
Using `self` would be perfect for Oxigraph.

Reply:
So we agree.

This raises another question, though:
since methods such as `bindings` and `graph` consume `self`, it does not make much sense for them to return an `Option`. If `None` is returned, the user can not "go back" and try something else...

Put it another way, in most (if not all) situations, the user will `unwrap` the returned option. Either they know in advance what kind of result they wil get:
```rust
    let res = rdf.query("ASK { :s :p :o: }");
    if res.boolean().unwrap() {
        // ...
    }
```
or they need to `match` on `res.kind()`
```rust
    let res = rdf.query(some_query_of_unknwon_kind);
    match res.kind() {
        Bindings => {
            handle_bindings(res.bindings().unwrap());
        }
        Graph => {
            handle_graph(res.graph().unwrap());
        }
        // ...
    }
```

So wouldn't it be more natural if all the kind-specific methods would return the expected value directly *or panic if the result has the wrong kind*? This is really an open question, I haven't made up my mind about it...

Reply:
Wouldn't it be possible to create different `SparqlResult` traits to represent results of the different query forms (say: `SparqlSelectResult`, `SparqlAskResult`, `SparqlConstructResult`)?

Perhaps they can all derive from a common `SparqlResult` trait, so that the return value of `SparqlDataset::query` is still a `SparqlResult` but in practice, you'll get a specific one that matches the associated type `SparqlDataset::Result`?

This way, you can only get, e.g., a `boolean` if your `SparqlResult` type also implements `SparqlAskResult`, and `SparqlAskResult::boolean` (or even `SparqlAskResult::result`) could return a `bool` rather than an `Option<bool>`.

I also wonder whether it might make more sense to replace the associated type `SparqlDataset::Result` with a generic type parameter to the `SparqlDataset::query` function, so that on the same representation of a dataset, you could do something like:

```rust
let sparql_result = my_dataset.query::<SparqlAskResult>(r"
    PREFIX foaf:    <http://xmlns.com/foaf/0.1/>
    ASK  { ?x foaf:name  "Alice" }
")?;
let bool_result = sparql_result.boolean(); // or maybe: sparql_result.result();
```

Reply:
@yever
Regarding your last proposal (type parameter in the `query` method), this would not be practical when the query is provided by the user or an external program, and the developer does not know in advance which kind of result to expect. So we need an enum at some level, on which programs can match. This is achieved by the `kind` method of my proposal.

Regarding your first proposal (different traits for the different kinds of results), it would indeed be cleaner. I was first reluctant, because it would require converting the concrete result type to a Sophia-specific enum -- by contrast, the current trait can be directly implemented by Oxigraph's [`QueryResult`](https://docs.rs/oxigraph/0.1.1/oxigraph/sparql/enum.QueryResults.html). But this conversion has probably a negligible cost compared to evaluating a SPARQL query, so... worth a try.

What do you think of #91?

Reply:
I really like #91.